### PR TITLE
Fix: back handler for Drawer

### PR DESCRIPTION
**Pull request by parasharrajat**
#### Issues 
- [x] We have to click twice on the LHN to go to any chat.
- [x] When we resize the page LHN and Report body is not correctly aligned.
- [x] Drawer does not open on Mobile Web.

### Fixed Issues

$ https://github.com/Expensify/App/issues/4211
$ https://github.com/Expensify/App/issues/5971

#### https://github.com/Expensify/App/issues/4211
1. Open LHN on Android.
2. navigate to a report.
3. Use the back button on the Android device.
4. You should land on LHN.
5. Use the back button on the Android device.
6. App should be closed.

#### https://github.com/Expensify/App/issues/5971
There are no clear instructions for testing it but I believe this PR will fix this issue due to the reason that we are using upgraded Drawer Lib. The new version has removed transformation from the View for the permanent drawer which means it should remain static.

1. Follow the reproduction steps mentioned on the issue.
2. Or, Try to resize the Browser window multiple times to observe that the layout does not break.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### CODE REVIEW

**Comment by parasharrajat** [comment]
One issue that I found is when we open the Workspace Settings route, it does not open the Card page at the click of the Card link. This is known to me and I am fixing it https://github.com/Expensify/App/pull/5131 here.

**Review by Beamanator — DISMISSED** [review]
Tests are looking great!

One weird Android thing I noticed: When viewing the app in the simulator, I have to press twice on any report to navigate to it. Once I navigate to a report I can close it and navigate to it again with one click, but if I close and want to navigate to a new report I have to click that report twice :( But it looks like that's also happening on `main` so not caused by this 👍 

Still can't merge b/c `n6` hold

**Comment by parasharrajat** [comment]
I looked at the diff between the versions. 

There are a couple of changes which can be directly the cause of our issues.
1. History for the Drawer is changed. # 1
2. There are a couple of changes for the drawer-type parmanant. 

I think the first can be fixed directly in our app. We are managing the state ourself and thus we could be wrong somewhere.

For 2, I am yet to analyze the changes and their effect.

**Comment by parasharrajat** [comment]
OK, @Beamanator I have to open another issue with R-navigation for 2 on the above list. It's clearly a regression from the above commit.

My suggestion
1. I think the work here for 4211 and 5027 is done. So those two milestone should be considered completed (Upwork).
2. Now we have two more issues which is blocking this PR.
3. I can try to fix the first one. A new milestone as this is a new issue.
4. For point 2, I have to open another issue and provide reproduction and cooperate with maintainers to fix that. This is additional work for me and you know how tedious it could be, so I request separate milestone for this as well.

We can hold the payment for all the above until this PR and other issues are fixed.

What do you suggest as the best path forward?

**Comment by parasharrajat** [comment]
Well. @Beamanator There are a few things that you are missing.
1. https://github.com/Expensify/App/issues/4211 => $1000 https://github.com/Expensify/App/issues/4211#issuecomment-907461021.
2. https://github.com/Expensify/App/issues/5027 => Last Priced at $1000.


But I am still investigating the 2. I am looking at the code to see if this could a bug with our configuration so I will complete the repro to compare.

For 1, I believe it's related as it was not happening earlier and only after these changes. I have checked the changes and the lib author has changes regarding drawer history which are causing our logic to fail.
https://github.com/Expensify/App/blob/550dfe6c66f04b0210628743554e70fee19f13d6/src/libs/Navigation/CustomActions.js#L75-L87

Let me debug it further.

**Review by roryabraham — CHANGES_REQUESTED** [review]
Found a severe regression in this PR. On mobile web, if you're on the homescreen you can't navigate to chat reports.

File: src/libs/onlyWeb/index.js
Comment (Beamanator) [inline src/libs/onlyWeb/index.js:5]: Instead of mentioning native platforms **here**, I'd mention the native-related issue in `index.native.js`

File: src/libs/onlyWeb/index.js
Comment (Beamanator) [inline src/libs/onlyWeb/index.js:2]: What about Desktop? Also I think it's better to add a more detailed explanation of WHY we're doing this, taking no shortcuts. Maybe we can name this utility something clearer since `onlyWeb` doesn't describe what it's doing

**Review by Beamanator — COMMENTED** [review]
@parasharrajat I updated the milestone rewards [here](https://github.com/Expensify/App/pull/5745#issuecomment-948027934), thanks for pointing out the price increases that i missed!

Since you're thinking this PR ALSO solves #5971 can you please provide test steps in the PR description so we know how to test that that issue was fixed?

> The only drawback is when you resize between Mobile size(under props.isSmallScreenWidth) and desktop size (above props.isSmallScreenWidth), this switch will cause the render of the Drawer

Can you show a video of what you mean here? I will try to test that out tomorrow to see what you mean, but I'm out a bit early today.

Other comments below

File: src/libs/onlyWeb/index.js
Comment (parasharrajat) [reply src/libs/onlyWeb/index.js:2]: Yeah. I put the working logic in but I definitely need help with this. can you suggest a better Naming or technique so that I can set that only for Web or Desktop. There is one more PR, that I am currently working on, has the same situation where we have to disable a prop only for web and desktop. After merging, I will refactor the code to use the generic utility. 

For now, we only need a utility for the web and desktop. Please suggest a good name.

File: src/libs/onlyWeb/index.js
Comment (parasharrajat) [reply src/libs/onlyWeb/index.js:5]: OK.

**Review by Beamanator — CHANGES_REQUESTED** [review]
Let's add some explanation to these, then I think we're good to go

**Review by roryabraham — CHANGES_REQUESTED** [review]

**Review by roryabraham — DISMISSED** [review]
Code LGTM, but I'll leave it to @Beamanator to test + merge

**Review by marcaaron — CHANGES_REQUESTED** [review]

**Review by marcaaron — CHANGES_REQUESTED** [review]
Sorry, I know this has been a long PR, but these changes seem to have caused a not so great regression with navigating to a report

1. Navigate to a report
2. Press back button in header
3. Search for a user
4. Select chat
5. Land on LHN instead of chat 💥

**Comment by parasharrajat** [comment]
cc: @marcaaron @Beamanator

**Review by marcaaron — DISMISSED** [review]
Nice working good now thanks!

**Comment by marcaaron** [comment]
All yours @Beamanator.

File: src/libs/Navigation/AppNavigator/DrawerNavigator/index.js
Comment (Beamanator) [inline src/libs/Navigation/AppNavigator/DrawerNavigator/index.js:11]: Hmm this comment doesn't make much sense to me, and I reading this wrong?

> On mobile web, the non legacy version of the DrawerNavigator breaks the layout on Desktop...

We're only using the legacy version when the screen is small (`props.isSmallScreenWidth` is true) which is only applicable:
1. On desktop & web when the window is sized small
2. On mobile web always

Right? So the dynamic legacy implementation is needed only for web (non-mobile) & desktop right?

**Review by Beamanator — COMMENTED** [review]
Overall the changes look great, the back button on android is working great 👍 and resizing the window in desktop & web have no regressions 👍 

The only 🤷  things are:
1. refreshing on web / desktop when you have a chat open, takes you back to the LHN
    - As mentioned in earlier comments, you're planning to fix this in another PR, right?
2. the comment below
    - If you disagree with me / I'm confused, no change is necessary :D

File: src/libs/Navigation/AppNavigator/DrawerNavigator/index.js
Comment (parasharrajat) [reply src/libs/Navigation/AppNavigator/DrawerNavigator/index.js:11]: There are two issues here. based on the drawer version.
Legacy implementation:
1. WEB (> 800): breaks the layout on resize 
2. Mobile Web: Works well.

Non-Legacy(modern) implementation
1. WEB (> 800): Works well.
2. Mobile Web: Drawer does not work, Closing and opening of a drawer do not work.


So use Legacy for M-web and Modern on Desktop.
On native use Modern for all cases.

**Comment by parasharrajat** [comment]
> refreshing on web / desktop when you have a chat open, takes you back to the LHN
> As mentioned in earlier comments, you're planning to fix this in another PR, right?

Unfortunately, we won't be solving the https://github.com/Expensify/App/issues/5027#issuecomment-951417341 anymore. But yeah It would be solved separately.

File: src/libs/Navigation/AppNavigator/DrawerNavigator/index.js
Comment (Beamanator) [reply src/libs/Navigation/AppNavigator/DrawerNavigator/index.js:11]: Just to be clear, when you mention WEB or Desktop by themselves, do you mean web (non-mobile web) AND desktop?

Ex:
- `**WEB** (> 800): breaks the layout on resize` - does this apply to Desktop too?
- `So use Legacy for M-web and Modern on **Desktop**.` - does "Desktop" apply to web (non mobile web) too?

Comment (parasharrajat) [reply src/libs/Navigation/AppNavigator/DrawerNavigator/index.js:11]: WEB > 800 also applies to Desktop and Desktop also applies to WEB.

File: src/libs/Navigation/AppNavigator/DrawerNavigator/index.js
Comment (Beamanator) [inline src/libs/Navigation/AppNavigator/DrawerNavigator/index.js:12]: What about this?

```suggestion
        // On mobile web, the non legacy (modern) version of the DrawerNavigator has issues opening and closing
        // the drawer, so we always use the legacy version.
        // On web & desktop, the legacy version is needed on small screen sizes and modern elsewhere, otherwise
        // the layout breaks when the screen is resized.
```

File: src/libs/Navigation/AppNavigator/DrawerNavigator/index.js
Comment (Beamanator) [reply src/libs/Navigation/AppNavigator/DrawerNavigator/index.js:11]: What do you think about this: https://github.com/Expensify/App/pull/5745/files#r739667981? Is it accurate?

Comment (parasharrajat) [reply src/libs/Navigation/AppNavigator/DrawerNavigator/index.js:11]: Let's get this merged quickly. Any thoughts @Beamanator ?

File: src/libs/Navigation/AppNavigator/DrawerNavigator/index.js
Comment (parasharrajat) [reply src/libs/Navigation/AppNavigator/DrawerNavigator/index.js:12]: Sorry, it seems, too wordy. Maybe we can shorten it a bit.

Comment (parasharrajat) [reply src/libs/Navigation/AppNavigator/DrawerNavigator/index.js:12]: On the mobile web, the non-legacy (modern) version of the DrawerNavigator has issues opening and closing, and the legacy version breaks the layout while resizing on Desktop.

Comment (parasharrajat) [reply src/libs/Navigation/AppNavigator/DrawerNavigator/index.js:12]: If you think this is fine, I will commit. Otherwise, let's go with yours. @Beamanator

Comment (Beamanator) [reply src/libs/Navigation/AppNavigator/DrawerNavigator/index.js:12]: I like it 👍 But at the end, please add "and web" so we know "the legacy version breaks the layout while resizing on desktop AND web"

**Comment by parasharrajat** [comment]
@Beamanator Comment Updated. Ready for merge.

**Review by Beamanator — APPROVED** [review]
Thanks for all of this 💪

File: package.json
Comment (kidroca) [inline package.json:96]: Hey @Beamanator @parasharrajat this change here seems to have broken the Attachment Picker for me on iOS: https://expensify.slack.com/archives/C01GTK53T8Q/p1635874753343600

Comment (parasharrajat) [reply package.json:96]: Hey, @kidroca this is PR is reverted now. Thanks for notifying me.

**Pull request merged**
